packages/contracts/contracts/optimistic-ethereum/OVM/bridge/tokens/Abs_L2DepositedToken.sol

```diff
@@ -139,66 +140,98 @@ abstract contract Abs_L2DepositedToken is iOVM_L2DepositedToken, OVM_CrossDomain
-     * @param _amount Amount of the token to withdraw
+     * @param _amount Amount of the token to withdraw.
+     * @param _data Data to forward to L1.
+     * @param _l1Gas Gas limit for the provided message.
      */
     function withdraw(
-        uint _amount
+        uint _amount,
+        bytes calldata _data,
+        uint32 _l1Gas
     )
         external
         override
         virtual
         onlyInitialized()
     {
-        _initiateWithdrawal(msg.sender, _amount);
+        _initiateWithdrawal(
+            msg.sender,
+            msg.sender,
+            _amount,
+            _data,
+            _l1Gas
+        );
     }
 
     /**
-     * @dev initiate a withdraw of some token to a recipient's account on L1
-     * @param _to L1 adress to credit the withdrawal to
-     * @param _amount Amount of the token to withdraw
+     * @dev initiate a withdraw of some token to a recipient's account on L1.
+     * @param _to L1 adress to credit the withdrawal to.
+     * @param _amount Amount of the token to withdraw.
+     * @param _data Data to forward to L1.
+     * @param _l1Gas Gas limit for the provided message.
      */
     function withdrawTo(
         address _to,
-        uint _amount
+        uint _amount,
+        bytes calldata _data,
+        uint32 _l1Gas
     )
         external
         override
         virtual
         onlyInitialized()
     {
-        _initiateWithdrawal(_to, _amount);
+        _initiateWithdrawal(
+            msg.sender,
+            _to,
+            _amount,
+            _data,
+            _l1Gas
+        );
     }
 
     /**
      * @dev Performs the logic for deposits by storing the token and informing the L2 token Gateway of the deposit.
      *
-     * @param _to Account to give the withdrawal to on L1
-     * @param _amount Amount of the token to withdraw
+     * @param _from Account to pull the deposit from on L2.
+     * @param _to Account to give the withdrawal to on L1.
+     * @param _amount Amount of the token to withdraw.
+     * @param _data Data to forward to L1.
+     * @param _l1Gas Gas limit for the provided message on L1.
      */
     function _initiateWithdrawal(
+        address _from,
         address _to,
-        uint _amount
+        uint _amount,
+        bytes calldata _data,
+        uint32 _l1Gas
     )
         internal
     {
         // Call our withdrawal accounting handler implemented by child contracts (usually a _burn)
         _handleInitiateWithdrawal(_to, _amount);
 
         // Construct calldata for l1TokenGateway.finalizeWithdrawal(_to, _amount)
-        bytes memory data = abi.encodeWithSelector(
+        bytes memory message = abi.encodeWithSelector(
             iOVM_L1TokenGateway.finalizeWithdrawal.selector,
+            _from,
             _to,
-            _amount
+            _amount,
+            _data
         );
 
+        // Prevent tokens stranded on other side by taking
```

Comment:
hmmmm... a nit here is that we may not be strictly preventing by letting the user specify; it might be larger but still too small.

actually, forgot about this, but just looked it up and we [do have a mitigation in place](https://github.com/ethereum-optimism/optimism/blob/52d9e6b73362a203bdae3fa0182bd6e7c9693b13/packages/contracts/contracts/optimistic-ethereum/OVM/bridge/messaging/OVM_L1CrossDomainMessenger.sol#L236) for upping the gas.

Reply:
Can we resolve this comment?